vm based providers: Introduce Provision manager

oshoval opened the pull request:
Provision manager determines which providers to rebuild and which to retag according
previous released `KUBEVIRTCI_TAG`.

**Motivation:**
1. Freeze versions unless changed, less prone for regressions,
improves reproducibility and consistency.
3. Faster provisioning.
4. Decouple collision domains, if one provider is problematic, we can still build other ones,
without blocking them.

**Flow:**

It checks which files changed versus the latest `KUBEVIRTCI_TAG`.
For each changed file, it find the matching rule and accumulate a list of targets to rebuild.

https://github.com/kubevirt/project-infra/pull/2600
Will be used as presubmit that checks dry running, and also allows to see what will happen during
publish.

Tested `git ls-files` as the "git change" to see all the files are covered by the rules.
Pman asserts that a changed file has a rule, so in case someone adds a file without rule
it will return error.

**Notes:**

* We have the k8s patch version bumping bot that will keep us updated when needed.
* If desired we can even skip presubmits using this mechanism in case the files of the PR doesn't affect
a specific provider.
* We can improve the rule list by creating some more specific rules for some of the files instead
for their whole directory but lets do it on follow PR.
* In case we want to use native golang for git operations, consider https://github.com/go-git/go-git/issues/717

oshoval commented:
/uncc @davidvossel

oshoval commented on publish.sh:
```diff
@@ -13,6 +14,26 @@ TARGET_KUBEVIRT_REPO="quay.io/kubevirt"
+      else
+          IMAGES_TO_RETAG+=("$key")
+      fi
+  done < <(echo "$json_result" | jq -r 'to_entries[] | "\(.key):\(.value)"')
```
@brianmcarey 

Is it fine that we install please jq on the bootstrap image?
I will create the PR if so

brianmcarey replied:
It should be there already - https://github.com/kubevirt/project-infra/blob/43ae34e031803dc4d47fba5518e3411e9e514fc6/images/bootstrap/Dockerfile#L41

oshoval replied:
Right just saw thanks

oshoval commented:
@brianmcarey @dhiller 

Basically ready to review, may update a few here and there
Thanks

dhiller commented on cluster-provision/gocli/cmd/PROVISION_MANAGER.md:
```diff
@@ -0,0 +1,49 @@
+
+## Related Files:
+1. `hack/pman/rules.txt` - the rules that determine which folder/file affects which provider.
+2. `hack/pman/force` - changing this file allows to enforce rebuild of all vm based providers.
```
Why not adding a flag to the pman command?

oshoval replied:
You need a way to change a file in your commit, which trigger the pre submit / post submit to rebuild all the targets.
So you can use a flag only manually, but not on CI, unless you have another lane which you will run manually, but it is prone to forget.

You also need that it will affect just the current PR, so the logic of pman itself is used here - if the file isn't changed it won't redo it, so it will affect only one time.

dhiller commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,417 @@
+
+func getKubevirtciTag() (string, error) {
+	const kubevirtciTagUrl = "https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirtci/latest?ignoreCache=1"
+	output, err := runCommand("curl", []string{kubevirtciTagUrl})
```
use go http client here.

```
resp, err := http.Get("https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirtci/latest?ignoreCache=1")
...
```

dhiller commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,417 @@
+		}
+	}
+
+	printRulesDB(rulesDB)
```
Also the output would maybe better suited to be a struct that contains the information extracted from the config, but best would be to have everything captured directly in the config file.

oshoval replied:
It is a map, we sort it and print it.

Maybe you mean related to the yaml structure ?
Note that it renders dynamically the config file, it is not the config file raw.
It globs the directories and apply the targets (both are dynamic information).
So even if so, it would need to marshal.

Ouput example for our config file:
If you still think we need to do something please elborate .

```
.gitignore : [none]                          
LICENSE : [none]                             
Makefile : [none]                            
OWNERS : [none]                              
cluster-provision : [1.24 1.25 1.26-centos9] 
cluster-provision/centos8/* : [1.24 1.25]    
cluster-provision/centos9/* : [1.26-centos9] 
cluster-provision/gocli/* : [1.24 1.25 1.26-centos9] 
cluster-provision/images/* : [none]          
cluster-provision/k8s : [1.24 1.25 1.26-centos9] 
cluster-provision/k8s/1.24/* : [1.24]        
cluster-provision/k8s/1.25/* : [1.25]        
cluster-provision/k8s/1.26-centos9/* : [1.26-centos9] 
cluster-provision/tools/* : [none]           
cluster-up : [1.24 1.25 1.26-centos9]        
cluster-up/cluster : [1.24 1.25 1.26-centos9] 
cluster-up/cluster/external : [none]         
cluster-up/cluster/k3d-1.25-sriov/* : [none] 
cluster-up/cluster/k3d/* : [none]            
cluster-up/cluster/k8s-1.24/* : [1.24]       
cluster-up/cluster/k8s-1.25/* : [1.25]       
cluster-up/cluster/k8s-1.26-centos9/* : [1.26-centos9] 
cluster-up/cluster/kind-1.23-sriov/* : [none] 
cluster-up/cluster/kind-1.23/* : [none]      
cluster-up/cluster/kind-1.25-vgpu/* : [none] 
cluster-up/cluster/kind/* : [none]           
cluster-up/cluster/local : [none]            
cluster-up/hack : [1.24 1.25 1.26-centos9]   
hack : [1.24 1.25 1.26-centos9]              
hack/kustomization/* : [1.24 1.25 1.26-centos9] 
hack/pman/* : [1.24 1.25 1.26-centos9]       
publish.sh : [1.24 1.25 1.26-centos9]    
```

oshoval commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,417 @@
+func buildRulesDB(input io.Reader, targets []string) (map[string][]string, error) {
+	rulesDB := make(map[string][]string)
+
+	scanner := bufio.NewScanner(input)
```
Can we please do this on a follow up ?
i prefer to have a first version, easier to do improvements in steps

oshoval replied:
Done

oshoval commented on cluster-provision/gocli/cmd/PROVISION_MANAGER.md:
```diff
@@ -0,0 +1,49 @@
+
+## Notes:
+1. If you need to enforce provision of all providers, run and commit this file:
+`curl -sL https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirtci/latest?ignoreCache=1 > hack/pman/force`
```
See please https://github.com/kubevirt/kubevirtci/pull/982#discussion_r1149186378
adding force can be done, but not instead, it can help locally only,
because we need a way to run force on CI, without changing the lane / running manually

oshoval replied:
Added `BYPASS_PMAN` flag to publish.sh
because there is no need to run pman if we want all targets

oshoval commented:
Updated a bit the unit tests

oshoval commented on cluster-provision/gocli/cmd/PROVISION_MANAGER.md:
```diff
@@ -0,0 +1,49 @@
+`--debug` - Run in debug mode.
+Output has the following sections:
+1. Run Parameters such as the tag comparing to, and the targets detected.
+2. The expanded rules (based on rules file).
```
You can already have a "dry-run" of pman itself here (simply when running as standalone)
https://github.com/kubevirt/project-infra/pull/2600/files#diff-ab108ff0d13755bd671fe2f0a2397f0c0290dc6655d2a03659b2cee3fa5958f7R580

It does print all what is going to be done + the info described above
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2600/rehearsal-check-provision-manager/1640314585237950464

If you want to just dry run, simply run it as standalone, and not as part of publish,
unless we want to add a dry-run flag to publish as well,
but note please that we do need to keep the debug as well, as it prints more debug info

oshoval commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,417 @@
+		Args:  cobra.ExactArgs(0),
+	}
+	provision.Flags().String("tag", "", "kubevirtci tag to compare to, default: fetch latest")
+	provision.Flags().String("rules", "hack/pman/rules.txt", "rules file, default: hack/pman/rules.txt")
```
For `rules` you are right, but for the Bool it doesnt (one line below the one you marked)
also for the tag it doens't, as it is `""` and even mean `fetch lastest`
so i am fixing for `rules`

oshoval commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,417 @@
+
+	params.debug, err = cmd.Flags().GetBool("debug")
+	if err != nil {
+		return parameters{}, err
```
About `nil` (b), using an empty struct because:
it is a struct, so i can't return nil
`cmd/provision_manager.go:123:10: cannot use nil as parameters value in return statement`
unless we make it a pointer, but i think it is overkill

oshoval replied:
about (a), Cobra already prints an informative error
for example use `--tag` without value gives `flag needs an argument: --tag`

oshoval commented:
Addressed some of the comments, (see inline)
fixed a wrong print

oshoval commented:
Using yaml instead of txt file for the rules

oshoval commented:
Fixed a small unit test
Updated commit message of the first commit

It is still ready for review, but moving to draft in order to not consume resources.

oshoval commented on hack/pman/rules.yaml:
```diff
@@ -0,0 +1,43 @@
+# exclude supports only one value per pattern
+exclude:
+  - "pattern": "cluster-provision/centos8/*"
+    "exclude": "1.26-centos9"
```
Would need to support multi value here (separated by ',' or a list).
Because a new 1.27 centos9 is going to be out soon
Same for "target" of "specific"

oshoval commented:
Rebased

oshoval commented:
Added support of multi targets for `specific` and `exclude` rules.
Added validation that the `path` of the rules is valid.

oshoval commented:
Addressed some comments
updated help string

oshoval commented:
Rebased and updated rules to take 1.27 is based on centos9

oshoval commented:
Can you please review again ? all comments either addressed or commented back

oshoval commented:
Rebased

brianmcarey commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,519 @@
+		}
+	}
+
+	if errorFound {
```
Could we gather the errors that occur ? and log them after this message? 
It could help with troubleshooting - what do you think?

oshoval replied:
This is exactly why we have it and what we are doing
https://github.com/kubevirt/kubevirtci/pull/982/commits/0a205f06aca723a8c48129d29207bb5a97548585#diff-7ed4bf7ea9633967a33dfedd8c17285c330d117c226a8a8b084225ba2518b59eR258

brianmcarey commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,519 @@
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != STATUS_OK {
```
Is this required? there is a check for errors on L285 already?

And there is a check for parsing the body below too?

oshoval replied:
This why it is more explicit, there could be other errors like 404 or so, letting the previous checks pass.

Once you got status 200 you know there is content, you use that content as git hash
The content is valid basically, there is no need to check what is the content.
If theoretically the content is bad the git command will fail, but again, never saw it happens
and once we check status 200 it is to see that the server said all good and we can expect output and a valid one

oshoval commented:
Rebased

dhiller reviewed (commented):
@oshoval I am really sorry that it's taking me so long, therefore I am just adding what I found so far. I'll do another review soon.

oshoval commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,519 @@
+// NewProvisionManagerCommand determines which providers should be rebuilt
+func NewProvisionManagerCommand() *cobra.Command {
+	provision := &cobra.Command{
+		Use:   "pman",
```
I am pro using abbrivations please if possible

oshoval commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,417 @@
+	provision := &cobra.Command{
+		Use:   "pman",
+		Short: "provision manager determines which providers should be rebuilt",
+		RunE:  provisionManager,
```
personally i never saw deep details on those commands and never looked on those, usually it is in the readme,
If you won't insist, i prefer having it on the other places mentioned in the comments

oshoval commented on cluster-provision/gocli/build/Dockerfile:
```diff
@@ -1,4 +1,8 @@
-FROM scratch
+FROM alpine:latest
+
+RUN apk add git
```
and then we will need to install docker in the image arent we ?
i think it is over kill, i prefer to keep this simple please

btw we can consider using go-git
there is a little issue with it, but theoretically we can live with it (i just less like the fact it is not explicit as the git itself)
https://github.com/go-git/go-git/issues/717
it doesnt detect Rename perfect, it make it looks like Delete and Add
it can work for us because the deleted file will act as deleted file (checks if a matching rule exists, if yes act, else continue without error that it didnt find it, because it is a deleted file)
and the added file will search a rule, act if found, error if not found

oshoval commented on cluster-provision/gocli/cmd/provision_manager.go:
```diff
@@ -0,0 +1,519 @@
+	Targets []string `yaml:"targets"`
+}
+
+type Config struct {
```
moved the little notes into the table in the readme, most was already there

dhiller replied:
Again, I think that the documentation should go along the fields of a struct. Reason for this is that updates tend to be forgotten if the docs are further away. Maybe a link from the README to the struct may replace the table then.

oshoval commented on hack/pman/rules.yaml:
```diff
@@ -0,0 +1,45 @@
+# note: if a folder ends with "*" it means it is recursive,
```
https://github.com/kubevirt/kubevirtci/pull/982#discussion_r1199746845

oshoval commented on cluster-provision/gocli/cmd/PROVISION_MANAGER.md:
```diff
@@ -0,0 +1,50 @@
+c. Check each of the `dirname`/* of the file and its parent directories (until `.`, not included).  
+If a match is found, the assicated rule is accumulated.
+Match must be found, unless the file is deleted,
+because we have rules which are regex, so once some files, are deleted also their rule will be gone.  
```
Done

oshoval commented:
Thanks
Addressed comments / answered inline

brianmcarey reviewed (commented):
/lgtm

dhiller reviewed (changes requested):
A couple of nits, but then it should be good to go.

oshoval commented on publish.sh:
```diff
@@ -84,14 +124,19 @@ function publish_alpine_container_disk() {
-  git config user.email "rmohr+kubebot@redhat.com"
+  if [ "$CI" == "true" ]; then
+    git config user.name "kubevirt-bot"
+    git config user.email "rmohr+kubebot@redhat.com"
```
It is not related to this PR, can we do it on a follow PR please ?

oshoval commented:
Addressed comments 
rebased

Thanks

dhiller reviewed (changes requested):
Sorry, one more nit found, then we are good to go.

oshoval commented:
rebased

oshoval commented:
addressed comments

oshoval commented:
Thanks for the review

oshoval commented:
It seems there is a bug with the push here
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/logs/publish-kubevirtci/1665867790357106688

It did build all the providers from a glance, but pushed only one
